src/prommis/uky/costing/ree_plant_capcost.py

```diff
@@ -2675,3 +2736,672 @@ def costing_upper_bound_eq(c, p):
+                value (NPV) calculations to; if not a costing block, b should
+                be a flowsheet block to attach parameters and variables to
+            discount_percentage: rate at which currency devalues over time;
+                alternatively, this is the required rate of return on investment.
+            plant_lifetime: length of operating period in years.
+            total_capital_cost: value for total capital cost; ignored if b is not None.
+            annual_operating_cost: value for total operating cost; ignored if b is not None.
+            annual_revenue: value for total revenue; ignored if b is not None.
+            cost_year: assumed project start year for costs, which is the basis for NPV results
+            capital_expenditure_percentages: a list of values that sum to 100
+                representing how capital costs are spread over a capital
+                expenditure period; for example, an input of [10, 60, 30] is parsed
+                as a 3-year period where capital costs are spread as 10% in year 1,
+                60% in year 2,and 30% in year 3. The capital period precedes
+                the operating period. Set to None to indicate no expenditure period.
+            capital_escalation_percentage: rate at which capital costs
+                escalate during the capital expenditure period. Set to 0 to indicate expenditure
+                is spread but there is no cost escalation.
+            operating_inflation_percentage: inflation rate for operating costs
+                during the operating period. Set to 0 to indicate no inflation.
+            revenue_inflation_percentage: inflation rate for revenue during the
+                operating period. Set to 0 to indicate no inflation.
+            royalty_charge_percentage_of_revenue: percentage of revenue charged
+                as royalties; ignored if royalty_expression is not None. Set to zero
+                to indicate no royalties are charged.
+            royalty_expression: set the value or expression to calculate royalties
+            debt_percentage_of_CAPEX: percentage of CAPEX financed by debt; ignored if
+                debt_expression is not None. Set to zero to indicate no loans are taken out on capital.
+            debt_expression: set the value or expression to calculate total debt
+            loan_interest_percentage: interest rate for loan repayment.
+            loan_repayment_period: length of repayment period in years.
+            capital_depreciation_declining_balance_percentage: factor to use for declining balance
+                depreciation. Set to 0 to indicate that capital does not depreciate over time.
+        """
+
+        # check if costing block was passed, and extract necessary components
+        # variables and constraints will be added to the costing block
+        if isinstance(b, FlowsheetCostingBlockData) and b.library == "REE":
```

Comment:
To help with testing and maintenance, I would suggest breaking all of this input verification out into separate methods that you can unit test.

E.g. all this code for testing the type of the `Block` should be a separate method so that you can quickly test this in isolation by throwing different types of blocks at it (dummied up with and without the expected arguments) so that you can fully test all possible cases without needing to make sure the rest of the code works with the dummy cases.